### DISCUSSION TITLE
P1 meter data to EVSE as MainsMet (API)

### DISCUSSION THREAD

**Discussion by Mvanmierlo** [serkri/SmartEVSE-3#202] [Q&A] [answered]
Hi 
Currently I have serkri 1.7.3 running and I managed to use Home Assistant to feed the MainsMet API with the currents of my P1 meter.
I am using a A20 based [P1 to ethernet device created by Marcel Zuidwijk ](https://www.zuidwijk.com/product/p1-reader-ethernet/) which is not based on an ESP (I have no wireless coverage at my P1 meter so I need ethernet connectivity.

With Home Assistant as an intermediate (I do not have the knowledge to have the A20 do something smart to feed the API of the EVSE with the proper info) I managed to get all current information to be send (and received) to my EVSE.

So far so good. BUT.
P1 telegrams do not show the direction of the current flow (from the grid or to the grid) I have solar so during the day my solar is feeding the grid most of the time.
I excepted this to be visible but it is not as the raw information :

_1-0:31.7.0(003*A)
1-0:51.7.0(001*A)
1-0:71.7.0(003*A)_

Is only providing in the current amperage that is flowing trough L1,L2 and L3 so it always shows up in the EVSE as consumption  although at this very moment I am not consuming but producing.

The only way to detect the direction of the power is using this part of the telegram : 
_1-0:21.7.0(00.000*kW)
1-0:41.7.0(00.000*kW)
1-0:61.7.0(00.000*kW)
1-0:22.7.0(00.649*kW)
1-0:42.7.0(00.312*kW)
1-0:62.7.0(00.872*kW)_

The 21,41,61 values being consumption **from the grid** and the 22,42,62 values being production **to the grind**
Challenge for me no lies that the EVSE needs Ampers and not kW's

I am not a programmer but my guess is I am not the only one that is wishing to use the P1 meter info to feed the EVSE MainsMet API so my question is get some help either by a feature request in the software to feed the P1 info direct to the API based on the kW values and have the firmware do the math in regard of consumption or production.

OR can some one help me with fixing this in Home Assistant / the E20 module or another way that I haven't come up with yet :-)

**Comment by dingo35** [comment]
This problem is addressed multiple times on tweakers.net, forum "zelfbouw laadpaal ervaringen".

**Reply by Mvanmierlo** [reply]
appologies for being the biggest noob here but could you be so kind to point me to the sollution ? 
Found people solving it with an ESP but as mentioned I am not using an ESP so your help would be greatly appriciated and appologies being the noob here.

**Comment by dingo35** [comment]
Im sorry Im as big a noob as you are on P1 interface devices, I use modbus kWh-meters.
Just keep on reading, Im pretty sure your use case is described and solved there.

**Reply by Mvanmierlo** [reply, marked as answer]
Found a solution (with help from the HA forum) so want to share this with you (and my guess is that more Home Assistant users will like this solution as this rules out the need of the sensor box.
Here's my code in my configuration.yaml of Home Assistante to do the trick :


```
#Data naar SMARTEVSe
shell_command:
  dsmrtosmartevse: 'curl -X POST "http://192.168.x.x/currents?L1={{ states("nettocurrent_l1") | float*10 }}&L2={{ states("nettocurrent_l2") | float*10 }}&L3={{ states("nettocurrent_l3") | float*10 }}" -H "accept: application/json" -H "Content-Type: application/json" -d {}'

template:
  - sensor:
      - name: "nettocurrent_l1"
        unit_of_measurement: A
        device_class: current
        state: >
          {% set solarl1 = states('sensor.electricity_meter_power_production_phase_l1') | fl
[configuration.yaml.txt](https://github.com/serkri/SmartEVSE-3/files/12697942/configuration.yaml.txt)
[automations.yaml.txt](https://github.com/serkri/SmartEVSE-3/files/12697944/automations.yaml.txt)
oat(0) %}
          {% set currentl1 = states('sensor.electricity_meter_current_phase_l1') | float(0) %}
          {{ currentl1 * -1 if solarl1 > 0 else currentl1 }}
        availability: >
          {{ states('sensor.electricity_meter_power_production_phase_l1') | is_number and 
            states('sensor.electricity_meter_current_phase_l1') | is_number }}
  - sensor:
      - name: "nettocurrent_l2"
        unit_of_measurement: A
        device_class: current
        state: >
          {% set solarl2 = states('sensor.electricity_meter_power_production_phase_l2') | float(0) %}
          {% set currentl2 = states('sensor.electricity_meter_current_phase_l2') | float(0) %}
          {{ currentl2 * -1 if solarl2 > 0 else currentl2 }}
        availability: >
          {{ states('sensor.electricity_meter_power_production_phase_l2') | is_number and 
            states('sensor.electricity_meter_current_phase_l2') | is_number }}
  - sensor:
      - name: "nettocurrent_l3"
        unit_of_measurement: A
        device_class: current
        state: >
          {% set solarl3 = states('sensor.electricity_meter_power_production_phase_l3') | float(0) %}
          {% set currentl3 = states('sensor.electricity_meter_current_phase_l3') | float(0) %}
          {{ currentl3 * -1 if solarl3 > 0 else currentl3 }}
        availability: >
          {{ states('sensor.electricity_meter_power_production_phase_l3') | is_number and 
            states('sensor.electricity_meter_current_phase_l3') | is_number }}
```

**Comment by dingo35** [comment]
Thx for your contribution; however github mangles the code.
Could you upload the file somewhere or post a Pull Request?

**Reply by dingo35** [reply]
You can post the file as attachment here...

**Comment by Mvanmierlo** [comment]
#Here are both files txt extension was necessary to enable uploading here yaml extension is not allowed to be uploaded

[configuration.yaml.txt](https://github.com/serkri/SmartEVSE-3/files/12697950/configuration.yaml.txt)
[automations.yaml.txt](https://github.com/serkri/SmartEVSE-3/files/12697951/automations.yaml.txt)

And [my first repository ](https://github.com/Mvanmierlo/P1-via-Ha-to-SmartEVSE-3) is a fact here you can find all details and files too.

**Reply by dingo35** [reply]
Thanks for your contribution, added as 6b754bd45bc3ece165115946e7b5b74235d594c0

**Comment by dingo35** [comment]
By the way, changed /3 to /1 to have the once per second update....
